source/distributed/full_distributed_tria.cc

```diff
@@ -0,0 +1,499 @@
+              {
+                if ((((child[binary_entry] >> (j * dim))) &
+                     (GeometryInfo<dim>::max_children_per_cell - 1)) !=
+                    (((parent[binary_entry] >> (j * dim))) &
+                     (GeometryInfo<dim>::max_children_per_cell - 1)))
+                  return false;
+
+
+                ++child_level;
+                if (child_level == n_child_indices)
+                  break;
+              }
+
+            ++binary_entry;
+          }
+        return true;
+      }
+    } // namespace internal
+
+    template <int dim, int spacedim>
+    void
+    Triangulation<dim, spacedim>::reinit(
+      const ConstructionData<dim, spacedim> &construction_data)
+    {
+      // check if there are locally relevant coarse-grid cells
+      if (construction_data.cells.empty() == true) // no
+        {
+          // 1) create a dummy hypercube
+          GridGenerator::hyper_cube(*this, 0, 1, false);
+
+          // 2) mark cell as artificial
+          for (auto cell = this->begin(); cell != this->end(); cell++)
+            {
+              cell->set_subdomain_id(dealii::numbers::artificial_subdomain_id);
+              cell->set_level_subdomain_id(
+                dealii::numbers::artificial_subdomain_id);
+            }
+
+          // 3) setup dummy mapping between locally relevant coarse-grid cells
+          //    and global cells
+          this->coarse_cell_id_to_coarse_cell_index_vector.emplace_back(0, 0);
+          this->coarse_cell_index_to_coarse_cell_id_vector.push_back(0);
+        }
+      else
+        {
+          // 1) store `coarse-cell index to coarse-cell id`-mapping
+          this->coarse_cell_index_to_coarse_cell_id_vector =
+            construction_data.coarse_cell_index_to_coarse_cell_id;
+
+
+
+          // 2) setup `coarse-cell id to coarse-cell index`-mapping
+          std::map<types::coarse_cell_id, unsigned int>
+            coarse_cell_id_to_coarse_cell_index_vector;
+          for (unsigned int i = 0;
+               i < construction_data.coarse_cell_index_to_coarse_cell_id.size();
+               i++)
+            coarse_cell_id_to_coarse_cell_index_vector
+              [construction_data.coarse_cell_index_to_coarse_cell_id[i]] = i;
+
+          for (auto i : coarse_cell_id_to_coarse_cell_index_vector)
+            this->coarse_cell_id_to_coarse_cell_index_vector.emplace_back(i);
+
+
+
+          // 3) create coarse grid
+          const SubCellData subcelldata;
+          dealii::parallel::Triangulation<dim, spacedim>::create_triangulation(
+            construction_data.vertices, construction_data.cells, subcelldata);
+
+
+
+          // 4) create all levels via a sequence of refinements
+          const auto &cell_infos = construction_data.cell_infos;
+          for (unsigned int ref_counter = 0; ref_counter < cell_infos.size();
+               ref_counter++)
+            {
+              // a) perform refinement
+              if (ref_counter > 0)
```

Comment:
This confused me a little. Is the code more complicated if you do the refinement for ref_counter==0 except the last?